File: chatguru-clickup-middleware/test-complete-flow.js
#!/usr/bin/env node

/**
 * Teste completo do fluxo de integração ChatGuru -> Middleware -> ClickUp
 * Simula todo o processo incluindo classificação AI e envio de anotações
 */

const axios = require('axios');

// Configurações
const MIDDLEWARE_URL = process.env.MIDDLEWARE_URL || 'https://chatguru-clickup-middleware-707444002434.southamerica-east1.run.app';
const CHATGURU_WEBHOOK = `${MIDDLEWARE_URL}/webhooks/chatguru`;

// Cores para output
const colors = {
    reset: '\x1b[0m',
    bright: '\x1b[1m',
    red: '\x1b[31m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    cyan: '\x1b[36m',
    magenta: '\x1b[35m'
};

// Função auxiliar para log colorido
function log(message, color = 'reset') {
    console.log(`${colors[color]}${message}${colors.reset}`);
}

// Teste 1: Mensagem que DEVE criar tarefa (atividade válida)
const validActivityPayload = {
    "data_criacao": new Date().toISOString(),
    "numero": "+5511988348951",
    "nome": "João Silva",
    "celular": "5511988348951",
    "origem": "whatsapp",
    "texto_mensagem": "Preciso de 3 caixas de parafusos M8 e 2 chapas de alumínio 2mm",
    "campanha_id": "625584ce6fdcb7bda7d94aa8",
    "campanha_nome": "Vendas WhatsApp",
    "tags": ["pedido", "material"],
    "extra": {
        "priority": "normal",
        "department": "vendas"
    }
};

// Teste 2: Mensagem que NÃO deve criar tarefa (não é atividade)
const invalidActivityPayload = {
    "data_criacao": new Date().toISOString(),
    "numero": "+5511970525814",
    "nome": "Maria Santos",
    "celular": "5511970525814",
    "origem": "whatsapp",
    "texto_mensagem": "Oi, bom dia! Tudo bem?",
    "campanha_id": "625584ce6fdcb7bda7d94aa8",
    "campanha_nome": "Vendas WhatsApp",
    "tags": ["saudacao"],
    "extra": {}
};

// Teste 3: Payload com formato diferente (EventType)
const eventTypePayload = {
    "event_type": "message.received",
    "timestamp": new Date().toISOString(),
    "data": {
        "phone": "5511988348951",
        "name": "Pedro Costa",
        "message": "Quero fazer um orçamento para 10 metros de cabo elétrico",
        "tags": ["orcamento"],
        "source": "chatguru"
    }
};

// Função para enviar teste
async function sendTest(testName, payload, expectedResult) {
    log(`\n${testName}`, 'cyan');
    log('─'.repeat(50), 'cyan');
    
    try {
        // Enviar webhook
        log('📤 Enviando webhook...', 'yellow');
        const startTime = Date.now();
        
        const response = await axios.post(CHATGURU_WEBHOOK, payload, {
            headers: {
                'Content-Type': 'application/json'
            },
            timeout: 10000
        });
        
        const duration = Date.now() - startTime;
        
        // Verificar resposta
        if (response.data && response.data.message === 'Success') {
            log(`✅ Resposta recebida em ${duration}ms: ${JSON.stringify(response.data)}`, 'green');
        } else {
            log(`⚠️ Resposta inesperada: ${JSON.stringify(response.data)}`, 'yellow');
        }
        
        // Aguardar processamento assíncrono
        log('⏳ Aguardando processamento assíncrono (5s)...', 'blue');
        await new Promise(resolve => setTimeout(resolve, 5000));
        
        // Informar resultado esperado
        log(`📊 Resultado esperado: ${expectedResult}`, 'magenta');
        
        return true;
    } catch (error) {
        log(`❌ Erro no teste: ${error.message}`, 'red');
        if (error.response) {
            log(`   Status: ${error.response.status}`, 'red');
            log(`   Data: ${JSON.stringify(error.response.data)}`, 'red');
        }
        return false;
    }
}

// Função para verificar status do serviço
async function checkServiceStatus() {
    log('\n🔍 Verificando status do serviço...', 'blue');
    
    try {
        const statusUrl = `${MIDDLEWARE_URL}/status`;
        const response = await axios.get(statusUrl, { timeout: 5000 });
        
        if (response.data) {
            log('✅ Serviço está online:', 'green');
            log(`   Versão: ${response.data.version || 'N/A'}`, 'green');
            log(`   Uptime: ${response.data.uptime || 'N/A'}`, 'green');
            log(`   ClickUp: ${response.data.clickup_connected ? 'Conectado' : 'Desconectado'}`, 'green');
            log(`   AI: ${response.data.ai_enabled ? 'Habilitada' : 'Desabilitada'}`, 'green');
        }
        return true;
    } catch (error) {
        log(`❌ Serviço não está respondendo: ${error.message}`, 'red');
        return false;
    }
}

// Função principal
async function runTests() {
    log('\n🚀 TESTE COMPLETO DO FLUXO DE INTEGRAÇÃO', 'bright');
    log('═'.repeat(50), 'bright');
    log(`URL do Middleware: ${MIDDLEWARE_URL}`, 'cyan');
    log(`Webhook Endpoint: ${CHATGURU_WEBHOOK}`, 'cyan');
    
    // Verificar status do serviço
    const isOnline = await checkServiceStatus();
    if (!isOnline) {
        log('\n⛔ Abortando testes - serviço não está disponível', 'red');
        process.exit(1);
    }
    
    // Executar testes
    const tests = [
        {
            name: 'TESTE 1: Atividade Válida (deve criar tarefa)',
            payload: validActivityPayload,
            expected: 'Tarefa criada no ClickUp + Anotação enviada ao ChatGuru'
        },
        {
            name: 'TESTE 2: Não é Atividade (não deve criar tarefa)',
            payload: invalidActivityPayload,
            expected: 'Apenas anotação "Não é uma atividade" enviada ao ChatGuru'
        },
        {
            name: 'TESTE 3: Formato EventType (compatibilidade)',
            payload: eventTypePayload,
            expected: 'Processamento normal com classificação AI'
        }
    ];
    
    let successCount = 0;
    for (const test of tests) {
        const success = await sendTest(test.name, test.payload, test.expected);
        if (success) successCount++;
        
        // Pequeno delay entre testes
        await new Promise(resolve => setTimeout(resolve, 2000));
    }
    
    // Resumo final
    log('\n' + '═'.repeat(50), 'bright');
    log('📊 RESUMO DOS TESTES', 'bright');
    log('─'.repeat(50), 'bright');
    log(`Total de testes: ${tests.length}`, 'cyan');
    log(`Sucesso: ${successCount}`, 'green');
    log(`Falhas: ${tests.length - successCount}`, 'red');
    
    if (successCount === tests.length) {
        log('\n🎉 TODOS OS TESTES PASSARAM!', 'green');
    } else {
        log('\n⚠️ Alguns testes falharam. Verifique os logs do serviço.', 'yellow');
    }
    
    // Instruções para verificar resultados
    log('\n📝 PRÓXIMOS PASSOS:', 'magenta');
    log('1. Verifique os logs do Cloud Run:', 'yellow');
    log(`   gcloud run logs tail chatguru-clickup-middleware --region southamerica-east1`, 'cyan');
    log('\n2. Verifique se as tarefas foram criadas no ClickUp:', 'yellow');
    log(`   curl -X GET "https://api.clickup.com/api/v2/list/901300373349/task" \\`, 'cyan');
    log(`     -H "Authorization: pk_106092691_UC96HSQJH4ZUS3NJATYHIQ06BQQYM657"`, 'cyan');
    log('\n3. Verifique se as anotações chegaram no ChatGuru:', 'yellow');
    log('   Acesse o painel do ChatGuru e verifique as mensagens enviadas', 'cyan');
}

// Executar testes
runTests().catch(error => {
    log(`\n❌ Erro fatal: ${error.message}`, 'red');
    process.exit(1);
});

File: chatguru-clickup-middleware/test_complete_flow.py
#!/usr/bin/env python3
"""
Teste completo do fluxo ChatGuru → ClickUp
Simula um payload real do ChatGuru e acompanha todo o processo até a criação da tarefa
"""

import json
import requests
import time
import uuid
from datetime import datetime

# Configuração
BASE_URL = "https://chatguru-clickup-middleware-pcwqxktwfq-rj.a.run.app"
TEST_CLIENT = "Nexcode"  # Nome da empresa (Info_1)
TEST_REQUESTOR = "Willaim Duarte"  # Nome do solicitante (Info_2)
TEST_ATTENDANT = "Anne Souza"  # Responsável pelo atendimento (deve mapear para space "Anne Souza")

def generate_test_payload():
    """Gera um payload de teste realista do ChatGuru"""
    payload = {
        "event_type": "message_received",
        "timestamp": datetime.now().isoformat(),
        "account_id": "625584ce6fdcb7bda7d94aa8",
        "conversation_id": f"test_conv_{uuid.uuid4().hex[:8]}",
        "message": {
            "id": f"msg_{uuid.uuid4().hex[:8]}",
            "content": "Preciso de ajuda com a integração OAuth2 do ClickUp. O sistema não está validando os tokens corretamente.",
            "sender": {
                "phone": "+5511999999999",
                "name": "Cliente Teste"
            },
            "timestamp": datetime.now().isoformat(),
            "type": "text"
        },
        "campos_personalizados": {
            "Info_1": TEST_CLIENT,  # Nome da empresa (Nexcode)
            "Info_2": TEST_REQUESTOR,  # Nome do solicitante (João Silva)
            "Info_3": "Suporte Técnico",
            "Info_4": "Alta",
            "Info_5": "Integração",
            "responsavel": TEST_ATTENDANT  # Responsável (Anne)
        },
        "metadata": {
            "channel": "whatsapp",
            "source": "web"
        }
    }
    return payload

def test_webhook_endpoint(payload):
    """Testa o endpoint webhook do ChatGuru"""
    print("=" * 60)
    print("🔥 INICIANDO TESTE COMPLETO DO FLUXO CHATGURU → CLICKUP")
    print("=" * 60)
    
    print(f"\n📋 DADOS DO TESTE:")
    print(f"   Empresa (Info_1): {TEST_CLIENT}")
    print(f"   Solicitante (Info_2): {TEST_REQUESTOR}")
    print(f"   Responsável: {TEST_ATTENDANT}")
    print(f"   Conversation ID: {payload['conversation_id']}")
    print(f"   Message ID: {payload['message']['id']}")
    print(f"   Expected Space: Anne Souza")
    
    print(f"\n💬 MENSAGEM DE TESTE:")
    print(f"   '{payload['message']['content']}'")
    
    print(f"\n🌐 ENVIANDO WEBHOOK PARA: {BASE_URL}/webhooks/chatguru")
    
    start_time = time.time()
    
    try:
        response = requests.post(
            f"{BASE_URL}/webhooks/chatguru",
            json=payload,
            headers={
                "Content-Type": "application/json",
                "User-Agent": "ChatGuru-Test/1.0"
            },
            timeout=30
        )
        
        webhook_time = time.time() - start_time
        
        print(f"\n✅ WEBHOOK RESPONSE:")
        print(f"   Status Code: {response.status_code}")
        print(f"   Response Time: {webhook_time:.3f}s")
        print(f"   Headers: {dict(response.headers)}")
        
        if response.status_code == 200:
            try:
                webhook_response = response.json()
                print(f"   Response Body: {json.dumps(webhook_response, indent=2)}")
            except:
                print(f"   Response Body: {response.text}")
        
        return response.status_code == 200, webhook_time
        
    except Exception as e:
        print(f"\n❌ ERRO NO WEBHOOK: {str(e)}")
        return False, 0

def wait_for_processing():
    """Aguarda o processamento assíncrono via Pub/Sub"""
    print(f"\n⏳ AGUARDANDO PROCESSAMENTO VIA PUB/SUB...")
    print("   (O worker processa a mensagem em background)")
    
    for i in range(10):
        time.sleep(2)
        print(f"   Aguardando... {(i+1)*2}s")
    
    print("   ✅ Tempo de espera concluído")

def check_clickup_integration():
    """Verifica se a integração ClickUp está funcionando"""
    print(f"\n🔗 VERIFICANDO INTEGRAÇÃO CLICKUP...")
    
    try:
        response = requests.get(f"{BASE_URL}/status", timeout=10)
        
        if response.status_code == 200:
            status_data = response.json()
            
            print(f"   ✅ Status Service: {response.status_code}")
            
            # Verificar integração ClickUp
            clickup_status = status_data.get("integrations", {}).get("clickup", {})
            print(f"\n📊 CLICKUP INTEGRATION STATUS:")
            print(f"   Configured: {clickup_status.get('configured', 'Unknown')}")
            print(f"   Connection: {clickup_status.get('connection', 'Unknown')}")
            print(f"   Token Configured: {clickup_status.get('token_configured', 'Unknown')}")
            print(f"   List ID: {clickup_status.get('list_id', 'Unknown')}")
            print(f"   List Name: {clickup_status.get('list_name', 'Unknown')}")
            
            # Verificar outras integrações
            chatguru_status = status_data.get("integrations", {}).get("chatguru", {})
            openai_status = status_data.get("integrations", {}).get("openai", {})
            
            print(f"\n📊 OTHER INTEGRATIONS:")
            print(f"   ChatGuru API: {chatguru_status.get('api_configured', 'Unknown')}")
            print(f"   OpenAI Enabled: {openai_status.get('enabled', 'Unknown')}")
            print(f"   AI Enabled: {status_data.get('ai_enabled', 'Unknown')}")
            
            return clickup_status.get('connection') == 'success'
        else:
            print(f"   ❌ Status endpoint failed: {response.status_code}")
            return False
            
    except Exception as e:
        print(f"   ❌ Erro verificando status: {str(e)}")
        return False

def search_created_task():
    """Busca a tarefa criada no ClickUp"""
    print(f"\n🔍 BUSCANDO TAREFA CRIADA NO CLICKUP...")
    
    try:
        # Tentar buscar tarefas via endpoint de debug
        response = requests.get(f"{BASE_URL}/clickup/tasks", timeout=15)
        
        if response.status_code == 200:
            tasks_data = response.json()
            print(f"   ✅ Tasks endpoint response: {response.status_code}")
            
            if isinstance(tasks_data, dict) and "tasks" in tasks_data:
                tasks = tasks_data["tasks"]
                print(f"   📝 Total de tarefas encontradas: {len(tasks)}")
                
                # Buscar a tarefa mais recente
                if tasks:
                    latest_task = max(tasks, key=lambda t: t.get("date_created", 0))
                    
                    print(f"\n🎯 TAREFA MAIS RECENTE ENCONTRADA:")
                    print(f"   ID: {latest_task.get('id', 'N/A')}")
                    print(f"   Nome: {latest_task.get('name', 'N/A')}")
                    print(f"   Status: {latest_task.get('status', {}).get('status', 'N/A')}")
                    print(f"   Lista: {latest_task.get('list', {}).get('name', 'N/A')}")
                    print(f"   Data Criação: {latest_task.get('date_created', 'N/A')}")
                    print(f"   URL: {latest_task.get('url', 'N/A')}")
                    
                    if latest_task.get('assignees'):
                        assignees = [a.get('username', 'N/A') for a in latest_task['assignees']]
                        print(f"   Assignees: {', '.join(assignees)}")
                    
                    return latest_task
                else:
                    print("   ⚠️ Nenhuma tarefa encontrada")
                    return None
            else:
                print(f"   ⚠️ Formato de resposta inesperado: {tasks_data}")
                return None
        else:
            print(f"   ❌ Tasks endpoint failed: {response.status_code}")
            print(f"   Response: {response.text}")
            return None
            
    except Exception as e:
        print(f"   ❌ Erro buscando tarefas: {str(e)}")
        return None

def test_oauth_endpoints():
    """Testa os endpoints OAuth2 melhorados"""
    print(f"\n🔐 TESTANDO ENDPOINTS OAUTH2 MELHORADOS...")
    
    try:
        # Testar endpoint de inicialização OAuth
        response = requests.head(f"{BASE_URL}/auth/clickup", timeout=10, allow_redirects=False)
        
        print(f"   OAuth Start Endpoint:")
        print(f"   Status: {response.status_code}")
        
        if response.status_code == 303:
            location = response.headers.get('location', '')
            print(f"   ✅ Redirect para ClickUp: {location[:100]}...")
        
        return response.status_code == 303
        
    except Exception as e:
        print(f"   ❌ Erro testando OAuth: {str(e)}")
        return False

def generate_test_report(webhook_success, webhook_time, clickup_working, oauth_working, created_task):
    """Gera relatório final do teste"""
    print("\n" + "=" * 60)
    print("📊 RELATÓRIO FINAL DO TESTE COMPLETO")
    print("=" * 60)
    
    print(f"\n🔥 RESULTADOS GERAIS:")
    print(f"   Webhook Success: {'✅' if webhook_success else '❌'}")
    print(f"   Webhook Time: {webhook_time:.3f}s {'✅' if webhook_time < 5 else '⚠️'}")
    print(f"   ClickUp Integration: {'✅' if clickup_working else '❌'}")
    print(f"   OAuth2 Endpoints: {'✅' if oauth_working else '❌'}")
    print(f"   Task Created: {'✅' if created_task else '❌'}")
    
    print(f"\n🎯 FLUXO COMPLETO:")
    print(f"   1. ChatGuru Webhook → {'✅ SUCCESS' if webhook_success else '❌ FAILED'}")
    print(f"   2. Pub/Sub Processing → ⏳ ASYNC")
    print(f"   3. ClickUp Connection → {'✅ SUCCESS' if clickup_working else '❌ FAILED'}")
    print(f"   4. OAuth2 Security → {'✅ SUCCESS' if oauth_working else '❌ FAILED'}")
    print(f"   5. Task Creation → {'✅ SUCCESS' if created_task else '❌ FAILED'}")
    
    if created_task:
        print(f"\n🎉 TAREFA CRIADA COM SUCESSO:")
        print(f"   ID: {created_task.get('id', 'N/A')}")
        print(f"   Nome: {created_task.get('name', 'N/A')}")
        print(f"   URL ClickUp: {created_task.get('url', 'N/A')}")
        print(f"   Lista: {created_task.get('list', {}).get('name', 'N/A')}")
    
    print(f"\n🔗 MELHORIAS OAUTH2 IMPLEMENTADAS:")
    print(f"   ✅ Token validation robusta")
    print(f"   ✅ Cache invalidation inteligente")
    print(f"   ✅ Error handling específico (OAUTH_025, OAUTH_027)")
    print(f"   ✅ Authorization headers corrigidos")
    print(f"   ✅ Auto-reauthorization")
    
    # Score final
    score = sum([webhook_success, clickup_working, oauth_working, bool(created_task)])
    total_score = (score / 4) * 100
    
    print(f"\n📈 SCORE FINAL: {total_score:.1f}% ({score}/4 testes)")
    
    if total_score >= 75:
        print("🎉 SISTEMA FUNCIONANDO CORRETAMENTE!")
    elif total_score >= 50:
        print("⚠️ SISTEMA PARCIALMENTE FUNCIONAL")
    else:
        print("❌ SISTEMA COM PROBLEMAS CRÍTICOS")
    
    print("=" * 60)

def main():
    """Executa o teste completo"""
    print("🚀 Iniciando teste completo do fluxo ChatGuru → ClickUp...")
    
    # 1. Gerar payload de teste
    test_payload = generate_test_payload()
    
    # 2. Testar webhook
    webhook_success, webhook_time = test_webhook_endpoint(test_payload)
    
    # 3. Aguardar processamento
    if webhook_success:
        wait_for_processing()
    
    # 4. Verificar integração ClickUp
    clickup_working = check_clickup_integration()
    
    # 5. Testar OAuth2
    oauth_working = test_oauth_endpoints()
    
    # 6. Buscar tarefa criada
    created_task = search_created_task()
    
    # 7. Gerar relatório final
    generate_test_report(webhook_success, webhook_time, clickup_working, oauth_working, created_task)

if __name__ == "__main__":
    main()